File: solutions/clojure/robot-name/1/src/robot_name.clj
(ns robot-name
    (:require [clojure.string :as str]))

(defn rand-char [start limit]
    (char (+ start (rand-int limit)))
)

(defn rand-letter []
    (rand-char 65 26)
)

(defn rand-digit []
    (rand-char 48 10)
)

(defn random-name [_]
    (str/join [(rand-letter) (rand-letter) (rand-digit) (rand-digit) (rand-digit)])
)

(defn robot []
  (atom (random-name ""))  
)

(defn robot-name [bot]
  @bot 
)

(defn reset-name [bot]
  (swap! bot random-name)
)

File: solutions/csharp/robot-name/1/RobotName.cs
using System;
using System.Linq;
using System.Collections.Generic;

public class Robot
{
    public Robot()
    {
        Reset();
    }
    public string Name {get; private set;}

    static Random rand = new Random();
    static HashSet<string> used_names = new HashSet<string>();

    public void Reset()
    {
        for(;;) {
            var candidate = string.Join("",
                rand.NextChars('A', 'Z').Take(2)
                .Concat(rand.NextChars('0', '9').Take(3))
            );

            if (used_names.Add(candidate))
            {
                Name = candidate;
                return;
            }
        }
    }
}

public static class RandomExt {

    public static IEnumerable<char> NextChars(this Random rand, char first, char last)
    {
        for (;;)
            yield return (char)rand.Next(first, last + 1);
    }
}